### PR TITLE
feat: added template option for new notes

### PR DESCRIPTION
**Pull request by MrMysterius**
Added Settings option and functionality for a template file with replacement placeholder {{title}} for the title with `#` and {{link}} for the link to the previous note. Also, logic to check and abort if neither one is present at least once.

Had to restructure / change the behavior of some parts to make the abort logic possible without letting the plugin make any changes to notes.

### CODE REVIEW

**Comment by MrMysterius** [comment]
@Eforen

**Comment by Eforen** [comment]
I will CR this soon. I am away from the computer right now.

**Comment by Eforen** [comment]
BTW thank you for tagging me! Best way to get my attention don't hesitate to tag me again if I don't have input for you by tomorrow night

**Comment by MrMysterius** [comment]
Merry Christmas and Quick Reminder for the Pull Request and the other one (#42) that is a bug fix @Eforen

Also, I think I found a small Issue with the title placeholder replacement, since I took a variable that was made somewhere before for the title there is now unnecessary whitespace and new lines added to the file. I'm not sure on the whole pull request stuff on GitHub. So I would say fix that after this is merged.

**Comment by Eforen** [comment]
Fix it pre merge just make a new commit to your branch and then the PR will update.

Tag me and let me know when you feel it's ready. I don't want anything merged that is not ready to be deployed to the end users.

File: src/main.ts [hunk 1]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
...
+      let template_content = "";
+      try {
+        template_content = await this.app.vault.adapter.read(this.settings.templateFile);
+      } catch (err) {
+        new Notice(`Couldn't read template file ${this.settings.templateFile} make sure that it is a valid file...`);
+        return;
+      }
+
+      if (!titleRegex.test(template_content) || !backlinkRegex.test(template_content)) {
+        new Notice("Title {{title}} or Backlink {{link}} placeholder in template missing. Please add and try again...");
```
Comment (Eforen) [inline src/main.ts:286]: Instead of just throwing a notice let's add a setting of if we want to throw an error or just silently continue with the default. Also I think the default should be to silently continue. Assume not everyone will have a template.

**Review by Eforen — CHANGES_REQUESTED** [review]
Please change this before we merge

File: src/main.ts (hunk 1 above)
Comment (MrMysterius) [reply src/main.ts:286]: I check beforehand ([This Line | Will Also Comment Separately](https://github.com/Dyldog/luhman-obsidian-plugin/pull/43/commits/2e1cf2b9a4aef2f137558a1f4247292f838226f5#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dR276)), if the template setting is empty, so basically saying you don't have a template selected or chosen to not use one, then it will go use the previous version of note creation ([Here | Will Comment As Well](https://github.com/Dyldog/luhman-obsidian-plugin/pull/43/commits/2e1cf2b9a4aef2f137558a1f4247292f838226f5#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dR293)), so notice will be triggered there ever.

Now If a template is set, non-empty setting, it will check first if the file is readable ([Here | Will Comment As Well](https://github.com/Dyldog/luhman-obsidian-plugin/pull/43/commits/2e1cf2b9a4aef2f137558a1f4247292f838226f5#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dR278)) and if not throw an error notice that is informing the user they might have a non-existent path to a template set or other file reading problems that might occur. After that, it will check, if the two placeholders that exist are present and if not, telling the user the template is malformed (The Code Marked for Change). Since without both of them or at least without the title, the file/note will be problematic afterwards for the search function since it won't guarantee the title is an H1 which is from my understanding used for the search/open command of the plugin.

So I would say this should stay because it will inform the user there is something wrong with the template they are using, if they are using a template at all. Though, I would say I should add a more detailed description to the setting informing the user that this is the case, and they are required to use both of the placeholders.

File: src/main.ts [hunk 2]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
     } else {
       titleContent = "";
     }
-    const fullContent = titleContent + content;
-    const file = await this.app.vault.create(path, fullContent);
+
+    let file = null;
+    const backlinkRegex = /{{link}}/g;
+    const titleRegex = /{{title}}/g;
+    if (this.settings.templateFile && this.settings.templateFile != "") {
```
Comment (MrMysterius) [inline src/main.ts:276]: This checks, if a template is even in use (Maybe need whitespace trimming)

File: src/main.ts [hunk 3]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
...
+
+      if (!titleRegex.test(template_content) || !backlinkRegex.test(template_content)) {
+        new Notice("Title {{title}} or Backlink {{link}} placeholder in template missing. Please add and try again...");
+        return;
+      }
+
+      const file_content = template_content.replace(titleRegex, titleContent).replace(backlinkRegex, fileLink);
+      file = await this.app.vault.create(path, file_content);
+      successCallback();
+    } else {
```
Comment (MrMysterius) [inline src/main.ts:293]: This is the previous file creation logic, that is used if no template is used.

File: src/main.ts [hunk 4]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
...
     }
-    const fullContent = titleContent + content;
-    const file = await this.app.vault.create(path, fullContent);
+
+    let file = null;
+    const backlinkRegex = /{{link}}/g;
+    const titleRegex = /{{title}}/g;
+    if (this.settings.templateFile && this.settings.templateFile != "") {
+      let template_content = "";
+      try {
```
Comment (MrMysterius) [inline src/main.ts:278]: Here it will check if the template file path is valid and throw an error if not readable (aka. wrong path, no file present, other problems while reading)

**Review by MrMysterius — COMMENTED** [review]
This is the previous file creation logic, that is used if no template is used.

File: src/main.ts [hunk 5]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
...
+        new Notice("Title {{title}} or Backlink {{link}} placeholder in template missing. Please add and try again...");
+        return;
+      }
+
+      const file_content = template_content.replace(titleRegex, titleContent).replace(backlinkRegex, fileLink);
+      file = await this.app.vault.create(path, file_content);
+      successCallback();
+    } else {
+      const fullContent = titleContent + fileLink;
+      file = await this.app.vault.create(path, fullContent);
```
Comment (MrMysterius) [inline src/main.ts:295]: This is the previous file creation logic, that is used if no template is used.

File: src/main.ts [hunk 6]
```diff
@@ -251,8 +269,33 @@ export default class NewZettel extends Plugin {
...
+
+    let file = null;
+    const backlinkRegex = /{{link}}/g;
+    const titleRegex = /{{title}}/g;
+    if (this.settings.templateFile && this.settings.templateFile != "") {
+      let template_content = "";
+      try {
+        template_content = await this.app.vault.adapter.read(this.settings.templateFile);
+      } catch (err) {
+        new Notice(`Couldn't read template file ${this.settings.templateFile} make sure that it is a valid file...`);
```
Comment (MrMysterius) [inline src/main.ts:281]: Here it will check if the template file path is valid and throw an error if not readable (aka. wrong path, no file present, other problems while reading)

File: src/main.ts (hunk 1 above)
Comment (MrMysterius) [reply src/main.ts:288]: @Eforen

File: src/main.ts (hunk 6 above)
Comment (electblake) [reply src/main.ts]: @Eforen

**Comment by electblake** [comment]
Maybe I'm not seeing it, but wouldn't it make more sense to support https://github.com/SilentVoid13/Templater ?

To be fair I haven't implemented templater in another plugin, but seems standard as most people have this plugin (or really should imho)

**Comment by MrMysterius** [comment]
> Maybe I'm not seeing it, but wouldn't it make more sense to support https://github.com/SilentVoid13/Templater ?
> 
> To be fair I haven't implemented templater in another plugin, but seems standard as most people have this plugin (or really should imho)

This is exactly what it would allow since you can set a file that has templater stuff in it. 

Thanks though you reminded me to get back an finish this up.

**Comment by MrMysterius** [comment]
This should be polished enough and working as expected.

What does this pull request **recapped**:

**Adds an option** into the **plugin setting** for setting a **template file**. This setting is a **text box**, for setting a **file path**.
If this setting **has something in** it that isn't just whitespace, **it will try to use this template** for new files, **if not** uses **old** file creation **logic**.
If there is **no template file at the given template file path**, it will **throw an error notice** informing the user no template file was found and **abort** the file creation.
When it finds a template file **it will try to apply the template**, it will **check** if **two placeholders** that need to be used are present. One is the **Title** and the other is the **Backlink**. It will **throw an error notice if not both of them are present** and **abort** the file creation.
The placeholders are as followed: Title is `{{title}}` and Backlink is `{{link}}`
They are applied and replaced globally, not just once.
The `{{title}}` is replaced by this: `# <title given through input>`
The `{{link}}` is replaced by this: `[[<name of the file creation was called from>]]`

Now the file **creation concludes** and place the use either when aborted on the same aka. not do anything or into the newly created file.

This should be all. **Also, the Error Notices should be there and always shown, since you have to make a change to the config to even get there and are there to inform the user about a settings mistake / wrongly setup template.**

@Eforen, I hope you can get to this now and would merch this feature.

**Comment by HaxSam** [comment]
bump

File: src/main.ts [hunk 7]
```diff
@@ -334,30 +382,17 @@ export default class NewZettel extends Plugin {
...
-          : selectionPos.anchor;
-        editor!.replaceRange(
-          " ".repeat(spaceBefore) + newLink(title) + " ".repeat(spaceAfter),
-          virtualAnchor,
-          virtualHead
-        );
-        this.makeNote(nextPath(title), title, fileLink, true, openNewFile);
+        const virtualAnchor = anchorCorrect ? selectionPos.anchor : selectionPos.head;
+        const virtualHead = anchorCorrect ? selectionPos.head : selectionPos.anchor;
+        // editor!.replaceRange(" ".repeat(spaceBefore) + newLink(title) + " ".repeat(spaceAfter), virtualAnchor, virtualHead);
```
Comment (Eforen) [inline src/main.ts:387]: Why was this moved into the call back? I could be miss remembering this but doesn't that call back call on the new file not the old one? If not please let me know I could very well be wrong and am trying to get through all the PRs before bed. (Aiming for a working next version release on or before friday)

File: src/main.ts [hunk 8]
```diff
@@ -240,19 +257,50 @@ export default class NewZettel extends Plugin {
...
+        );
+        return;
+      }
+
+      if (!titleRegex.test(template_content) || !backlinkRegex.test(template_content)) {
+        new Notice("[LUHMAN] Template Malformed. Missing {{title}} and/or {{link}} placeholder. Please add them to the template and try again...", 15000);
+        return;
+      }
+
+      const file_content = template_content.replace(titleRegex, titleContent).replace(backlinkRegex, fileLink);
```
Comment (Eforen) [inline src/main.ts:295]: This will clober 41ca033 can you please unclober that for me if not I will try to get to this some time this week but I am not 100% I can do that as I have a lot of stuff this week with easter coming up and all.

Comment (Eforen) [reply src/main.ts:295]: Sorry I Accepted that one before looking at this one I probably would have had @electblake unclober instead if I had seen this one first but whats done is done.

File: src/main.ts (hunk 5 above)
Comment (Eforen) [reply src/main.ts:300]: This is where the primary clober will be but not sure how the new stuff for the aliases will interface with your stuff if at all in the template version but if I remember correctly templater does not work correctly in the frontmatter. (crying emoji)

File: src/main.ts [hunk 9]
```diff
@@ -334,30 +382,17 @@ export default class NewZettel extends Plugin {
...
-          ? selectionPos.head
-          : selectionPos.anchor;
-        editor!.replaceRange(
-          " ".repeat(spaceBefore) + newLink(title) + " ".repeat(spaceAfter),
-          virtualAnchor,
-          virtualHead
-        );
-        this.makeNote(nextPath(title), title, fileLink, true, openNewFile);
+        const virtualAnchor = anchorCorrect ? selectionPos.anchor : selectionPos.head;
+        const virtualHead = anchorCorrect ? selectionPos.head : selectionPos.anchor;
```
Comment (Eforen) [inline src/main.ts:386]: I like the compression of these 2 lines good job!

File: src/main.ts [hunk 10]
```diff
@@ -334,30 +382,17 @@ export default class NewZettel extends Plugin {
...
+        // editor!.replaceRange(" ".repeat(spaceBefore) + newLink(title) + " ".repeat(spaceAfter), virtualAnchor, virtualHead);
+        this.makeNote(nextPath(title), title, fileLink, true, openNewFile, () => {
+          editor!.replaceRange(" ".repeat(spaceBefore) + newLink(title) + " ".repeat(spaceAfter), virtualAnchor, virtualHead);
```
Comment (Eforen) [inline src/main.ts:389]: This seems less readable compressed...

**Review by Eforen — CHANGES_REQUESTED** [review]
I am sorry for making yours the cloberer I didn't check this one before the other so didn't see I should accept them in the opposet order >.<

File: src/main.ts (hunk 7 above)
Comment (MrMysterius) [reply src/main.ts:387]: It was moved into a callback because it placed replace a selection/placed with a link to the new note, but if were to be aborted because of wrong setting etc. it shouldn't make a link to a non-existing file aka. edit the old one.

File: src/main.ts (hunk 10 above)
Comment (MrMysterius) [reply src/main.ts:389]: Mmhh... don't know how to make it more readable, since I just copied the original replace method into a callback method. Could declare the callback method beforehand, but I don't think that makes stuff more readable necessarily.

File: src/main.ts (hunk 9 above)
Comment (MrMysterius) [reply src/main.ts:386]: I think that was my formatter at the time. I tried no to mess with the formatting of the original code, but it was near a part that was changed by me, so I probably committed the formatting as well. But if it's okay, then no problem, I guess. :)

File: src/main.ts (hunk 8 above)
Comment (MrMysterius) [reply src/main.ts:295]: Manual Merging on my end and moving / fixing stuff there.

**Comment by MrMysterius** [comment]
Hey @Eforen, It should all be working now with the changes from @electblake for the title alias. At least my tests did not point out any error or bugs. Manually merged the changes of the title alias into mine and moved some stuff to get here.

**Comment by MrMysterius** [comment]
Just a friendly reminder @Eforen

**Review by Eforen — APPROVED** [review]
LGTM

**Comment by Eforen** [comment]
Small change of plans looks like I can't work on this until Saturday I am gona try to get a build out the door around then though.

If you want to help out, one thing that would be really helpful is to collect all the PR numbers into a next build ticket so I don't have to when I ready.

**Comment by MrMysterius** [comment]
I'll see what I can do, probably will create an issue on the topic if I get to putting all the PR numbers together.

**Comment by Eforen** [comment]
> I'll see what I can do, probably will create an issue on the topic if I get to putting all the PR numbers together.

Awesome make sure to tag me if you do.

**Comment by Eforen** [comment]
I am going to change it to not include the `#` in the `{{title}}` key because this will allow wider usage. I will also add some verbage to make it more clear that the keys are case and space sensitive.

**Comment by Eforen** [comment]
I am also going to make both optional
